src/Metadata/Resource/Factory/ResourceMetadataFactoryTrait.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace ApiPlatform\Core\Metadata\Resource\Factory;
+
+use ApiPlatform\Core\Exception\ResourceClassNotFoundException;
+use ApiPlatform\Core\Metadata\Resource\Operation;
+use ApiPlatform\Core\Metadata\Resource\PaginationMetadata;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use ApiPlatform\Core\Annotation\Resource;
+
+trait ResourceMetadataFactoryTrait
+{
+    /**
+     * Returns the metadata from the decorated factory if available or throws an exception.
+     *
+     * @param ResourceMetadata|null $parentPropertyMetadata
+     * @param string                $resourceClass
+     *
+     * @throws ResourceClassNotFoundException
+     *
+     * @return ResourceMetadata
+     */
+    private function handleNotFound(ResourceMetadata $parentPropertyMetadata = null, string $resourceClass) : ResourceMetadata
+    {
+        if (null !== $parentPropertyMetadata) {
+            return $parentPropertyMetadata;
+        }
+
+        throw new ResourceClassNotFoundException(sprintf('Resource "%s" not found.', $resourceClass));
+    }
+
+    /**
+     * Compute ResourceMetadata operations.
+     *
+     * @param resource              $annotation
+     * @param ResourceMetadata|null $parentResourceMetadata
+     *
+     * @return ResourceMetadata
+     */
+    private function createMetadata(Resource $annotation, ResourceMetadata $parentResourceMetadata = null) : ResourceMetadata
+    {
+        if (!$parentResourceMetadata) {
+            return new ResourceMetadata(
+                $annotation->shortName,
+                $annotation->description,
+                $annotation->iri,
+                $annotation->itemOperations,
+                $annotation->collectionOperations,
+                $annotation->attributes
+            );
+        }
+
+        $resourceMetadata = $parentResourceMetadata;
+        foreach (['shortName', 'description', 'iri', 'attributes'] as $property) {
+            $resourceMetadata = $this->createWith($resourceMetadata, $property, $annotation->$property);
+        }
+
+        $resourceMetadata = $this->createWith($resourceMetadata, 'collectionOperations', $this->createOperations($annotation->collectionOperations));
+        $resourceMetadata = $this->createWith($resourceMetadata, 'itemOperations', $this->createOperations($annotation->itemOperations));
+
+        return $resourceMetadata;
+    }
+
+    /**
+     * Creates operation and pagination metadata from annotations.
+     *
+     * @param array|null $operationAnnotations
+     *
+     * @return array|null
+     */
+    private function createOperations(array $operationAnnotations = null)
+    {
+        if (null === $operationAnnotations) {
+            return [];
+        }
+
+        $operations = [];
+        //@TODO - Parse annotations/configuration and build OperationResourceMetadataFactory manually
```

Comment:
Here, I'd love to see an Operation merge instead of replacing defaults. Dunno if it's too much or maybe wrong but I'd allow less configuration for adding a custom operation.